src/components/search/InputWithSuggester.tsx

```diff
@@ -135,50 +138,90 @@ const InputWithSuggester: FC<InputWithSuggesterProps> = ({
-      const currentState: ParameterState = getComponentState(store.getState());
-      dispatch(fetchSuggesterOptions(createSuggesterParamFrom(currentState)))
-        .unwrap()
-        .then((data) => {
-          const options: OptionType[] = [];
-
-          const categorySuggestions = new Set<string>(
-            data.category_suggestions
-          );
-          const phrasesSuggestions = new Set<string>(
-            data.record_suggestions.suggest_phrases
-          );
-
-          // Get the intersection of categorySuggestions and phrasesSuggestions
-          const commonSuggestions = [...categorySuggestions].filter((item) => {
-            return phrasesSuggestions.has(item);
-          });
-
-          commonSuggestions.forEach((common: string) => {
-            // Remove the commonSuggestions from categorySuggestions and phrasesSuggestions
-            categorySuggestions.delete(common);
-            phrasesSuggestions.delete(common);
-
-            options.push({ text: common, group: OptionGroup.COMMON });
-          });
-
-          categorySuggestions.forEach((category: string) => {
-            options.push({ text: category, group: OptionGroup.CATEGORY });
-          });
-          phrasesSuggestions.forEach((phrase: string) => {
-            options.push({ text: phrase, group: OptionGroup.PHRASE });
+  const refreshOptions = useCallback(
+    async (inputValue: string) => {
+      // setPendingSearch to true to prevent doing search before refreshing options is finished
+      setPendingSearch(true);
+      try {
+        const currentState: ParameterState = getComponentState(
+          store.getState()
+        );
+        dispatch(fetchSuggesterOptions(createSuggesterParamFrom(currentState)))
+          .unwrap()
+          .then((data) => {
+            const categorySuggestions = new Set<string>(
+              data.category_suggestions
+            );
+            const phrasesSuggestions = new Set<string>(
+              data.record_suggestions.suggest_phrases
+            );
+
+            // Get the intersection of categorySuggestions and phrasesSuggestions
+            const commonSuggestions = [...categorySuggestions].filter(
+              (item) => {
+                return phrasesSuggestions.has(item);
+              }
+            );
+
+            // check if the current inputValue is a common key or not, then dispatch updateCommonKey
+            if (
+              commonSuggestions.some(
+                (item) => item.toLowerCase() === inputValue.toLowerCase()
+              )
+            ) {
+              dispatch(updateCommonKey(inputValue));
+            } else {
+              dispatch(updateCommonKey(""));
+            }
+
+            // Create array of all unique suggestions
+            const allSuggestions = [
+              ...new Set([
+                ...commonSuggestions,
+                ...categorySuggestions,
+                ...phrasesSuggestions,
+              ]),
+            ];
```

Comment:
if you extracting method, you don't need to combined actions here, it can be just a `Set`, and leave the converting task for different object.

```javascript
  // Create array of all unique suggestions
  const allSuggestions = new Set([
    ...commonSuggestions,
    ...categorySuggestions,
    ...phrasesSuggestions
  ]);
```

when you pass it to the extracted method, pass as `[...allSuggestions]`, It's OK, as it works and behaves exactly the same whether breaking it down or not. I'd try to avoid combining responsibilities/actions by a single object, that will be helpful when debugging.

Reply:
Yes I leave it here the reduce the duplication, but not in the extracted function

Reply:
you meant no changes or you will convert on the fly when passing to the method?

e.g

```javascript
const sortedSuggestions = sortByRelevance([...allSuggestions], inputValue);
```

Reply:
you are current using

```javascript
const sortedSuggestions = sortByRelevance(allSuggestions, inputValue);
```

because you already converted `sortedSuggestions` to `list` type while generating values to it as a `Set`.

Reply:
since I have spread the Set here 
`            const allSuggestions = [
              ...new Set([
                ...commonSuggestions,
                ...categorySuggestions,
                ...phrasesSuggestions,
              ]),
            ];
`
therefore I pass `allSuggestions` in `sortByRelevance` directly.
what you mentioned is a good point: keep allSuggestions as a Set, and modify the extracted function like:
`function sortByRelevance(items: string[] | Set<string>, target: string): string[] {
  const itemsArray = Array.from(items);
  // ... rest of the sorting logic
}
`